borg/archive.py

```diff
@@ -467,23 +467,40 @@ def rename(self, name):
         self.cache.chunk_decref(self.id, self.stats)
         del self.manifest.archives[self.name]
 
-    def delete(self, stats, progress=False):
-        unpacker = msgpack.Unpacker(use_list=False)
-        items_ids = self.metadata[b'items']
-        pi = ProgressIndicatorPercent(total=len(items_ids), msg="Decrementing references %3.0f%%", same_line=True)
-        for (i, (items_id, data)) in enumerate(zip(items_ids, self.repository.get_many(items_ids))):
+    def delete(self, stats, progress=False, forced=False):
+        error = False
+        try:
+            unpacker = msgpack.Unpacker(use_list=False)
+            items_ids = self.metadata[b'items']
+            pi = ProgressIndicatorPercent(total=len(items_ids), msg="Decrementing references %3.0f%%", same_line=True)
+            for (i, (items_id, data)) in enumerate(zip(items_ids, self.repository.get_many(items_ids))):
+                if progress:
+                    pi.show(i)
+                unpacker.feed(self.key.decrypt(items_id, data))
+                self.cache.chunk_decref(items_id, stats)
+                for item in unpacker:
+                    if b'chunks' in item:
+                        for chunk_id, size, csize in item[b'chunks']:
+                            self.cache.chunk_decref(chunk_id, stats)
             if progress:
-                pi.show(i)
-            unpacker.feed(self.key.decrypt(items_id, data))
-            self.cache.chunk_decref(items_id, stats)
-            for item in unpacker:
-                if b'chunks' in item:
-                    for chunk_id, size, csize in item[b'chunks']:
-                        self.cache.chunk_decref(chunk_id, stats)
-        if progress:
-            pi.finish()
-        self.cache.chunk_decref(self.id, stats)
+                pi.finish()
+        except Exception:
+            if not forced:
+                raise
+            error = True
```

Comment:
I think we can do a bit better and avoid the diaper by safeguarding the unpacker with a try/except msgpack.Xyz and the for chunk_... in item loop as well as the item decref with a separate try/except. So even if the first chunk would somehow get lost the following chunks should still be decref'd / deleted. (While a msgpack error wouldn't try any further).